src/devices/bus/sg1000_exp/sp400.cpp

```diff
@@ -0,0 +1,249 @@
+
+    port A connected to the panel buttons
+        "By analyzing the disassembly I found out that three pins (PA5,6,7 of the 6805) are used to set
+        different configurations. The resolution can be changed, probably to support different mechanics.
+        The baud rate is 4800."
+    port B connected to the motors
+    port C connected to the pen and misc signals
+    port D connected to M5L8035LP parallel data
+*/
+
+// this is currently incomplete: missing paper output and pen sensor, everything else is working
+
+#include "emu.h"
+#include "sp400.h"
+
+#include "cpu/mcs48/mcs48.h"
+#include "cpu/m6805/m68705.h"
+
+namespace {
+
+class sp400_printer_device : public device_t,
+	public device_sk1100_printer_port_interface
+{
+public:
+	sp400_printer_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock);
+
+	static constexpr feature_type unemulated_features() { return feature::PRINTER; }
+
+	virtual void device_add_mconfig(machine_config &config) override ATTR_COLD;
+	virtual void device_start() override ATTR_COLD;
+	virtual void device_reset() override ATTR_COLD;
+	virtual void device_reset_after_children() override ATTR_COLD;
+
+	const tiny_rom_entry * device_rom_region() const override ATTR_COLD;
+	virtual ioport_constructor device_input_ports() const override ATTR_COLD;
+
+	virtual void input_data (int state) override { m_data = state; }
+	virtual int output_busy () override { return m_busy; }
+
+private:
+	void dser_map(address_map &map);
+	void dser_p1_w(uint8_t data);
+	uint8_t dser_p2_r();
+	void dser_p2_w(uint8_t data);
+
+	uint8_t mot_pa_r();
+	void mot_pb_w(uint8_t data);
+	uint8_t mot_pc_r();
+	void mot_pc_w(uint8_t data);
+	uint8_t mot_pd_r();
+	void update_pen_state(u8 data);
+	u8 pen_sensor() { return 0; }  // not currently hooked up
+
+	required_device<i8035_device> m_dsercpu;     // "deserializer cpu"
+	required_device<m6805_hmos_device> m_motcpu; // "motor cpu"
+
+	int m_data, m_busy;
+	int m_dserdataout, m_dserstrobe;
+	int m_motbusy, m_motPenUp, m_motPenDown;
+	int m_pendown;
+
+	required_ioport m_buttons;
+};
+
+sp400_printer_device::sp400_printer_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock):
+	device_t(mconfig, SP400_PRINTER, tag, owner, clock),
+	device_sk1100_printer_port_interface(mconfig, *this),
+	m_dsercpu(*this, "dsercpu"),
+	m_motcpu(*this, "motcpu"),
+	m_data(0),
+	m_busy(1),
+	m_dserdataout(0xFF),
+	m_dserstrobe(1),
+	m_motbusy(0),
+	m_motPenUp(1),
+	m_motPenDown(1),
+	m_pendown(0),
+	m_buttons(*this, "BUTTONS")
+{
+}
+
+void sp400_printer_device::device_start()
+{
+	save_item(NAME(m_data));
+	save_item(NAME(m_busy));
+	save_item(NAME(m_dserdataout));
+	save_item(NAME(m_dserstrobe));
+	save_item(NAME(m_motbusy));
+	save_item(NAME(m_motPenUp));
+	save_item(NAME(m_motPenDown));
+	save_item(NAME(m_pendown));
+}
+
+void sp400_printer_device::device_reset()
+{
+}
+
+void sp400_printer_device::device_reset_after_children()
+{
+}
```

Comment:
You can leave out these – it’s not like reset propagates over a serial port anyway.